README.markdown
# cache_method

It's like `alias_method`, but it's `cache_method`!

Lets you cache the results of calling methods given their arguments. Like memoization, but stored in Memcached, Redis, etc. so that the cached results can be shared between processes and hosts.

## Real-world usage

<p><a href="http://brighterplanet.com"><img src="https://s3.amazonaws.com/static.brighterplanet.com/assets/logos/flush-left/inline/green/rasterized/brighter_planet-160-transparent.png" alt="Brighter Planet logo"/></a></p>

We use `cache_method` for [data science at Brighter Planet](http://brighterplanet.com/research) and in production at

* [Brighter Planet's impact estimate web service](http://impact.brighterplanet.com)
* [Brighter Planet's reference data web service](http://data.brighterplanet.com)

## Rationale

* It should be easy to cache instance methods
* It should be easy to cache methods that depend on object state
* It should be easy to uncache a method without clearing the whole cache
* It should be easy to do all that using a default in-process store, memcached, dalli (if you're on heroku), redis, etc. (all supported by this gem through the [cache gem](https://rubygems.org/gems/cache))

## Example

    require 'cache_method'
    class Blog

      attr_reader :name, :url

      def initialize(name, url)
        @name = name
        @url = url
      end

      def entries(date)
        # ...
      end

      # cache that slow method!
      cache_method :entries

      def update(stuff)
        # ...
      end

      # automatically clear cache for #entries when #update is called...
      cache_method_clear_on :update, :entries

      # custom cache key - not always required!
      def as_cache_key
        { :name => name, :url => url }
      end
    end

Then you can do

    my_blog.entries(Date.today) => first time won't be cached
    my_blog.entries(Date.today) => second time will come from cache

And clear them too

    my_blog.cache_method_clear :entries

(which doesn't delete the rest of your cache)

## Configuration (and supported cache clients)

By default, an in-process, non-shared cache is used.

You can set where the cache will be stored:

    CacheMethod.config.storage = Memcached.new '127.0.0.1:11211'

or

    CacheMethod.config.storage = Redis.new

or this might even work...

    CacheMethod.config.storage = Rails.cache

See `Config` for the full list of supported caches.

## Cache keys

Caching a method involves getting cache keys for

1. the object where the method is defined - for example, `my_blog.as_cache_key`
2. the arguments passed to the method - for example, `Marshal.dump(Date.today)`, because `Date#as_cache_key` is not defined

Then the cache keys are SHA-1 hashed and combined for an overall key:

    method signature + obj digest                                                                           + args digest
    Blog#entries     + SHA1(Marshal.dump({:name="Seamus's blog",:url=>"http://numbers.brighterplanet.com"}) + SHA1(Marshal.dump(Date.today))

Technically the full cache key is

    # when caching class methods
    method signature + generation + args digest
    # when caching instance methods
    method signature + obj digest + generation + args digest

(see "Generational caching" below for an explanation of the generation part)

### #as_cache_key

As above, you can define a custom cache key for an object:

    class Blog
      # [...]
      def as_cache_key
        { :name => name, :url => url }
      end
      # [...]
    end

If you don't define `#as_cache_key`, the default is to `Marshal.dump` the whole object. (That's not as terrible as it seems - marshalling is fast!)

### #to_cache_key (danger!)

There's another way to define a cache key, but it should be used with caution because it gives you total control.

The key is to make sure your `#to_cache_key` method identifies both the class and the instance!

### Comparison

<table>
  <tr>
    <th>Method</th>
    <th>Must uniquely identify class</th>
    <th>Must uniquely identify instance</th>
  </tr>
  <tr>
    <td><code>#as_cache_key</code></td>
    <td>N&dagger;</td>
    <td>Y</td>
  </tr>
  <tr>
    <td><code>#to_cache_key</code></td>
    <td>Y</td>
    <td>Y</td>
  </tr>
</table>

&dagger; The class name is automatically inserted for you by calling `object.class.name`, which is what causes all the trouble with `ActiveRecord::Associations::CollectionProxy`, etc.

## ActiveRecord

If you're caching methods on instances of `ActiveRecord::Base`, and/or using them as arguments to other cached methods, then you should probably define something like:

    class ActiveRecord::Base
      def as_cache_key
        attributes
      end
    end

If you find yourself passing association proxies as arguments to cached methods, this might be helpful:

    # For use in ActiveRecord 3.0.x
    class ActiveRecord::Associations::AssociationCollection
      # rare use of to_cache_key
      def to_cache_key
        [
          'ActiveRecord::Associations::AssociationCollection',
          proxy_owner.class.name,
          proxy_owner.id,
          proxy_reflection.name,
          conditions
        ].join('/')
      end
    end

    # For use in ActiveRecord 3.2.x
    class ActiveRecord::Associations::CollectionProxy
      # rare use of to_cache_key
      def to_cache_key
        owner = proxy_association.owner
        [
          'ActiveRecord::Associations::CollectionProxy',   # [included because we're using #to_cache_key instead of #as_cache_key ]
          owner.class.name,                                # User
          owner.id,                                        # 'seamusabshere'
          proxy_association.reflection.name,               # :comments
          scoped.where_sql                                 # "WHERE `comments`.`user_id` = 'seamusabshere'" [maybe a little bit redundant, but play it safe]
        ].join('/')
      end
    end

Otherwise, `cache_method` will call `user.comments.class.name` which causes the proxy to load the target, i.e. load all the Comment objects into memory. You probably don't want to load 1000 AR objects just to generate a cache key.

## Generational caching

Generational caching allows clearing the cached results for only one method, for example

    my_blog.cache_method_clear :entries

You can disable it to get a little speed boost

    CacheMethod.config.generational = false

Generational caching stores a separate generation key that never expires by default. To set a default TTL for these

    CacheMethod.config.default_generational_ttl = 120

## Debug

CacheMethod can warn you if your obj or args cache keys are suspiciously long.

    require 'cache_method'
    require 'cache_method/debug'

Then watch your logs.

## Module methods

You can put `#cache_method` right into your module declarations:

    module MyModule
      def my_module_method(args)
        # [...]
      end
      cache_method :my_module_method
    end

    class Tiger
      extend MyModule
    end
    
    class Lion
      extend MyModule
    end
    
Rest assured that `Tiger.my_module_method` and `Lion.my_module_method` will be cached correctly and separately. This, on the other hand, won't work:

    class Tiger
      extend MyModule
      # wrong - will raise NameError Exception: undefined method `my_module_method' for class `Tiger'
      # cache_method :my_module_method
    end

## Contributors

* [Seamus Abshere](https://github.com/seamusabshere)
* [Rubem Nakamura](https://github.com/rubemz)

## Copyright

Copyright 2012 Seamus Abshere


CHANGELOG
0.3.0 / 2022-10-16

* Enhancements

  * Ruby 3.x support

* Breaking changes

  * Cache key format changed to include kwargs; existing caches will be invalidated on upgrade

0.2.7 / 2015-04-03

* Enhancements

  * CacheMethod.config.default_generational_ttl so that generations can expire - thanks @itspriddle https://github.com/seamusabshere/cache_method/pull/10

0.2.6 / 2013-05-13

* Enhancements

  * cache_method_clear_on - for when a cache should be cleared whenever another method is called - thanks @rubemz ! https://github.com/seamusabshere/cache_method/pull/4
  * hypothetically more useful output when you require 'cache_method/debug'

0.2.5 / 2012-04-24

* Enhancements

  * If you immediately obtain the fetch mutex, don't double-get because for all you know you're the first setter
  * Better debug logging (require 'cache_method/debug')

0.2.4 / 2012-04-17

* Enhancements

  * Refactored to simplify and make thread-safe

0.2.3 / 2012-04-06

* Enhancements

  * Make it possible to work with proxies by defining #to_cache_key which---unlike #as_cache_key---does not cause #class to be run

0.2.2 / 2012-03-02

* Enhancements

  * Class name is now force-prepended to the result of #as_cache_key. This prevents objects of different classes that have the same #as_cache_key from being confused with each other. In any case, this sort of duck typing (?) doesn't seem appropriate for cache keys.
  * Enable "debug mode" by requiring "cache_method/debug"... it helps you find cache keys that are suspiciously long (and therefore slow to digest).

0.2.1 / 2012-03-01

* Enhancements

  * Use user-defined #as_cache_key for both obj and args digests (cache keys); previously it was only used for obj

0.2.0 / 2012-02-29

* API changes:

  * #clear_method_cache -> #cache_method_clear
  * #method_cache_hash -> #as_cache_key (see README)
  * doesn't rely on #hash at all any more (http://numbers.brighterplanet.com/2012/02/29/beware-of-string-hash-being-used-for-cache-keys/)
  * doesn't try to mitigate differences between Ruby 1.8 and 1.9 splats (http://www.ruby-forum.com/topic/98106)

* Enhancements

  * clearer method naming, I think
  * safer determination of cache keys - previously, in effect, args.map(&:to_s) was being used
  * no more autoload
  * remove direct dependency on activesupport
  * tests now run with dalli, green on MRI 1.8, MRI 1.9, and jruby 1.6.7

0.1.7 / yanked!

* Bug: used args.map(&:hash) instead of args.map(&:to_s)... which was actually worse because #hash is different across processes.

0.1.6 / November 17, 2011

* Last release before CHANGELOG in place


lib/cache_method/config.rb
require 'cache'

module CacheMethod
  # Here's where you set config options.
  #
  # Example:
  #     CacheMethod.config.storage = Memcached.new '127.0.0.1:11211'
  #     CacheMethod.config.default_ttl = 120 # seconds
  #
  # You'd probably put this in your Rails config/initializers, for example.
  class Config
    attr_reader :storages

    def initialize
      @mutex = ::Mutex.new
      @storages = StorageMap.new
    end

    # Whether to use "generational" caching. Default is true.
    #
    # Pro: enables clearing/flushing/expiring specific methods
    # Con: requires an extra trip to memcached to get the current "generation"
    #
    # Set to false if you just flush everything and don't need to selectively flush particular methods
    def generational=(boolean)
      @generational = boolean
    end

    def generational? #:nodoc:
      @generational == true or @generational.nil?
    end

    # Storage for the cache.
    #
    # Supported memcached clients:
    # * memcached[https://github.com/fauna/memcached] (either a Memcached or a Memcached::Rails)
    # * dalli[https://github.com/mperham/dalli] (either a Dalli::Client or an ActiveSupport::Cache::DalliStore)
    # * memcache-client[https://github.com/mperham/memcache-client] (MemCache, the one commonly used by Rails)
    #
    # Supported Redis clients:
    # * redis[https://github.com/ezmobius/redis-rb]
    #
    # Uses the cache[https://github.com/seamusabshere/cache] gem to wrap these, so support depends on that gem
    #
    # Example:
    #     CacheMethod.config.storage = Memcached.new '127.0.0.1:11211'
    def storage=(storage = nil)
      @storages[:default] = storage
    end

    def storage #:nodoc:
      @storages[:default] || @mutex.synchronize do
        @storages[:default] ||= ::Cache.new
      end
    end

    alias_method :storage=, :default_storage=
    alias_method :storage,  :default_storage

    # TTL for method caches. Defaults to 24 hours or 86,400 seconds.
    #
    # Example:
    #     CacheMethod.config.default_ttl = 120 # seconds
    def default_ttl=(seconds)
      @default_ttl = seconds
    end

    def default_ttl #:nodoc:
      @default_ttl || 86_400
    end

    # TTL for method generational caches. Defaults to 0 (never).
    #
    # Example:
    #     CacheMethod.config.default_generational_ttl = 120 # seconds
    def default_generational_ttl=(seconds)
      @default_generational_ttl = seconds
    end

    def default_generational_ttl #:nodoc:
      @default_generational_ttl || 0
    end
  end

  class StorageMap < Hash
    # Wraps the process of adding / changing a cache storage
    #
    # Examnple:
    #     CacheMethod.config.storages[:memcached] = Memcached.new '127.0.0.1:11211'
    def []=(name, storage)
      self[name] = ::Cache.wrap storage
    end
  end
end
